### PR TITLE
osd: allow multi dedicated journals for containers

### PR DESCRIPTION
**Pull request by leseb**
Fixes: https://github.com/ceph/ceph-ansible/issues/1729 and https://bugzilla.redhat.com/show_bug.cgi?id=1484466
Signed-off-by: Sébastien Han <seb@redhat.com>

### CODE REVIEW

File: roles/ceph-osd/templates/ceph-osd-run.sh.j2 [hunk 1]
```diff
@@ -44,8 +44,8 @@ create_dev_list $1
   -v /etc/localtime:/etc/localtime:ro \
   --device=/dev/${1} \
   --device=/dev/${1}1 \
-  {% if raw_journal_devices|length > 0 -%}
-  -e OSD_JOURNAL={{ raw_journal_devices[0] }} \
+  {% if raw_journal_devices -%}
+  -e OSD_JOURNAL={{ raw_journal_devices|random }} \
```
Comment (leseb) [inline roles/ceph-osd/templates/ceph-osd-run.sh.j2]: Note for myself, we need a random but predictable value here. We don't want to get a new value each time we run Ansible...

Comment (leseb) [inline roles/ceph-osd/templates/ceph-osd-run.sh.j2]: shuffle(seed=inventory_hostname), only works for Ansible`2.3.1`...

File: roles/ceph-osd/tasks/docker/start_docker_osd.yml [hunk 2]
```diff
@@ -16,6 +16,17 @@
   with_items: "{{ ceph_osd_docker_devices }}"
   register: osd_prepared
 
+- set_fact:
+    docker_env_args: -e OSD_JOURNAL={{ item.1 }}
```
Comment (bengland2) [inline roles/ceph-osd/tasks/docker/start_docker_osd.yml]: From Deepthi Dharwar's review:  The reference to item.1 seems out of scope.  Doesn't this var have to be referenced from a task that has a **with_together** clause?  It seems to be used in the tasks below that do have a with_together clause, and you could just add dedicated_devices|shuffle to the with_together  and reference it that way, right?

File: roles/ceph-osd/templates/ceph-osd-run.sh.j2 [hunk 3]
```diff
@@ -44,8 +44,8 @@ create_dev_list $1
   -v /etc/localtime:/etc/localtime:ro \
   --device=/dev/${1} \
   --device=/dev/${1}1 \
-  {% if dedicated_devices|length > 0 -%}
-  -e OSD_JOURNAL={{ dedicated_devices[0] }} \
+  {% if osd_scenario == 'non-collocated' -%}
+  -e OSD_JOURNAL={{ raw_journal_devices|shuffle(seed=inventory_hostname) }} \
```
Comment (Deepthidharwar) [inline roles/ceph-osd/templates/ceph-osd-run.sh.j2]: This has to be dedicated-devices and not raw-journal-devices.

File: roles/ceph-osd/tasks/docker/start_docker_osd.yml [hunk 4]
```diff
@@ -24,46 +35,22 @@
...
     {{ ceph_osd_docker_prepare_env }} \
-    "{{ ceph_docker_registry }}/{{ ceph_docker_image }}:{{ ceph_docker_image_tag }}" \
+    {{ ceph_docker_registry }}/{{ ceph_docker_image }}:{{ ceph_docker_image_tag }}
   with_together:
-    - "{{ ceph_osd_docker_devices }}"
+    - "{{ devices }}"
     - "{{ osd_prepared.results }}"
   when:
-    - item.1.get("rc", 0) != 0
+    - item.2.get("rc", 0) != 0
```
Comment (Deepthidharwar) [inline roles/ceph-osd/tasks/docker/start_docker_osd.yml]: Given that there is only devices and osd_prepared.results,  it should be item.1.get() instead of item.2.get()

**Comment by leseb** [comment]
luminous-ansible2.3-bluestore_docker_dmcrypt_journal_collocation needs https://github.com/ceph/ceph-docker/pull/670

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-update_docker_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-update_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-update_docker_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-update_docker_cluster
